### DISCUSSION TITLE
I have to stop using Powergird.

### DISCUSSION THREAD

**Discussion by JF-626164** [Power-Components/livewire-powergrid#236] [General]
With CDN being down last Friday, my company has decided to stop using Powergird because of how heavily it relies on cdn package calls. I am upset with this decision but with cdn crashing its shown us that we cannot rely on it to always be up.

**Comment by luanfreitasdev** [comment]
This is bad, as the company would have to develop its own logic to do what powergrid is doing. Everything in the config can be assigned 'null' (without quotes) and should work normally.

Powergrid has grown a lot recently and we are about to release the [Rules](https://github.com/Power-Components/livewire-powergrid/pull/217) feature which will be very important to the community.

**Comment by luanfreitasdev** [comment]
But can you reverse this?

The package takes the CDN that is in the config only when it is not null.

Put null for all of them and import via npm

![image](https://user-images.githubusercontent.com/33601626/150013483-0a9de3fa-b560-4aca-b139-1f419c23d2e4.png)

**Reply by JF-626164** [reply]
Where would this code go?

**Comment by luanfreitasdev** [comment]
config: livewire-powergrid.php

![image](https://user-images.githubusercontent.com/33601626/150013829-3184ab0e-912b-46b3-a4f0-be732b3110d7.png)

**Reply by JF-626164** [reply]
I'm arguing for it. I did add in null like you suggested. If I can get the flat_pickr working without having to call a cdn then we're golden. But I've been fighting with the code all day to get it to do that.

**Reply by luanfreitasdev** [reply]
I'm going to make a branch with this

**Comment by luanfreitasdev** [comment]
https://github.com/Power-Components/powergrid-demo/tree/bs5-webpack

Any doubt I am available

**Reply by luanfreitasdev** [reply]
can you give feedback?

**Reply by JF-626164** [reply]
I just read through the code, it looks good to me.

**Reply by luanfreitasdev** [reply]
Hi, did you make it?

**Reply by JF-626164** [reply]
Hey, I see the changes you made in the demo but I can't pull that branch onto my machine. I've tried "composer require power-components/powergird-demo:dev-bs5-webpack" but I'm getting errors saying "Could not find matching version."

**Reply by luanfreitasdev** [reply]
Hello, what I did was not in the package, it was in the demo, you can even download it and use the references I used in your project

Link: https://github.com/Power-Components/powergrid-demo/tree/bs5-webpack

Click: Download ZIP

![image](https://user-images.githubusercontent.com/33601626/151573782-76faef36-7c96-4f51-ad43-dad4e0a36e13.png)

thank

**Reply by JF-626164** [reply]
Okay, I've mimic'd most of the changes you made in that banch on my local. I'm getting a bunch of "GET http://localhost/null net::ERR_ABORTED 404 (Not Found)", its attempting to get a 'null' script.

**Reply by luanfreitasdev** [reply]
You add only these files in your project

**package.json**
```js
"devDependencies": {
        "@popperjs/core": "^2.11.2",
        "autoprefixer": "^10.2.5",
        "axios": "^0.21",
        "bootstrap": "^5.0.1",
        "laravel-mix": "^6.0.6",
        "lodash": "^4.17.19",
        "postcss": "^8.2.13",
        "resolve-url-loader": "^5.0.0",
        "sass": "^1.48.0",
        "sass-loader": "^12.4.0"
    },
    "dependencies": {
        "alpinejs": "^3.7.1",
        "bootstrap-select": "^1.14.0-beta2",
        "flatpickr": "^4.6.9",
        "jquery": "^3.6.0",
        "popper.js": "^1.16.1"
    }
```

**livewire-powergrid.php**
```php
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Template
    |--------------------------------------------------------------------------
    |
    | You will be able to know which theme pattern will be loaded
    | bootstrap or tailwind
    |
    */
    'theme' => 'bootstrap',

    /*
    |--------------------------------------------------------------------------
    | Plugins
    |--------------------------------------------------------------------------
    |
    | Plugins used: bootstrap-select when bootstrap, flatpicker.js to datepicker
    |
    */
    'plugins' => [
        /*
         * https://github.com/snapappointments/bootstrap-select
         */
        'bootstrap-select' => [
            'js'  => null,
            'css' => null
        ],
        /*
         * https://flatpickr.js.org
         */
        'flat_piker' => [
            'js'        => null,
            'css'       => null,
            'translate' => null,
            'locales'   => [
                'pt_BR' => [
                    'locale'     => 'pt',
                    'dateFormat' => 'd/m/Y H:i',
                    'enableTime' => true,
                    'time_24hr'  => true
                ]
            ]
        ]
    ],

    /*
    |--------------------------------------------------------------------------
    | Filters
    |--------------------------------------------------------------------------
    |
    | You choose which type of filter you want to use inline
    | to filter inside the table or outside the table
    | 'inline' or 'outside'
    |
    */

    'filter' => 'inline',

    /*
    |--------------------------------------------------------------------------
    | Cache
    |--------------------------------------------------------------------------
    |
    | When the data is cached, the search is much faster.
    | It is updated whenever the page is reloaded or a field is changed
    |
    */
    'cached_data' => true,

    /*
    |--------------------------------------------------------------------------
    | AlpineJS CDN
    |--------------------------------------------------------------------------
    |
    | Define here the CDN source for imported AlpineJS
    |
    */

    'alpinejs_cdn' => null,

     /*
    |--------------------------------------------------------------------------
    | Notification latest version
    |--------------------------------------------------------------------------
    |
    | Add the package: `composer require composer/composer --dev` to your project.
    | and change this value to `true`
    |
    */
    'check_version' => true
];
```

**app.js**
```javascript
window.Popper = require('popper.js').default;
window.$ = window.jQuery = require('jquery');

window.bootstrap = require('bootstrap');

import Alpine from 'alpinejs'
import flatpickr from "flatpickr";

window.Alpine = Alpine
window.flatpickr = flatpickr

require("./../../node_modules/flatpickr/dist/flatpickr.min.css");
require('./../../node_modules/bootstrap-select/dist/js/bootstrap-select');
require("./../../node_modules/bootstrap-select/dist/css/bootstrap-select.min.css");

import './../../vendor/power-components/livewire-powergrid/dist/powergrid'

Alpine.start()
```

**Reply by JF-626164** [reply]
Excellent! Thank you kindly for your help in sorting this out. I've modified my code to use the above and so far as I can tell its functioning prefectly.

**Reply by JF-626164** [reply]
Revisiting this after just over a week, the Powergrid table I'm using no longer functions like it used too. So the page loads, the table appears, the various built in sorts (I have a boolean toggle on the first column then two date pickers elsewhere) function like they ought too when clicked on, however the table does not refresh with new data. It doesn't pass the sort to the backend. The export does not work either when I click on it. I'm not sure how to fix this so I figured I'd let yall know about it. 

In the livewire-powergird.config everything is set to what the above config file has. Same for the app.js.

power-components/livewire-powergird: v2.2.5
AlpineJs: v3.8.1
bootstrap-select: v1.14.0-beta2
bootstrap: v5.1.3

**Reply by luanfreitasdev** [reply]
Nothing works ?. If you can share the class structure or make a video showing what is not working, then it will be easier to help.

**Reply by JF-626164** [reply]
What file would I get the class info? I don't think I'll be able to get a video.

**Reply by JF-626164** [reply]
I slept on the issue and went at it first thing when I got to work, I've got it working now. The only difference package wise between what I posted yesterday and what I'm currently using now is "bootstrap-select": "^1.13.18", rather than bootstrap-select: v1.14.0-beta2. 

I also moved the app.js code to the bottom of the file instead of the top. 

I'm also using the old config file still but all the cdns set to null.
